File: BusinessLogic/src/main/java/name/benjaminabbitt/wwwwwh/url/BusinessLogic/Bookmarks/StateMachine/NotCreatedBookmark.java
package name.benjaminabbitt.wwwwwh.url.BusinessLogic.Bookmarks.StateMachine;

import com.google.protobuf.Any;
import com.google.protobuf.Timestamp;
import name.benjaminabbitt.wwwwwh.url.BusinessLogic.Bookmarks.Bookmark;
import name.benjaminabbitt.evented.bookmarks.Bookmarks;
import name.benjaminabbitt.evented.core.Evented;

import java.util.stream.Stream;

public class NotCreatedBookmark implements Bookmark {
    private String name;
    private String url;


    @Override
    public Bookmark handle(Bookmarks.BookmarkDeleted bookmarkDeleted) throws InvalidState {
        throw new InvalidState();
    }

    @Override
    public Stream<Evented.EventPage.Builder> handle(Bookmarks.DeleteBookmark deleteBookmark) throws InvalidState {
        throw new InvalidState();
    }

    @Override
    public Bookmark handle(Bookmarks.BookmarkCreated bookmarkCreated) {
        this.url = bookmarkCreated.getUrl();
        this.name = bookmarkCreated.getName();
        return new ExtantBookmark(this);
    }

    @Override
    public Stream<Evented.EventPage.Builder> handle(Bookmarks.CreateBookmark createBookmark) {
        return Stream.of(Evented.EventPage.newBuilder()
                .setCreatedAt(Timestamp.newBuilder().build())
                .setSynchronous(false)
                .setEvent(Any.pack(
                        Bookmarks.BookmarkCreated.newBuilder()
                                .setName(createBookmark.getName())
                                .setUrl(createBookmark.getUrl())
                                .build()
                )));
    }

    public String getName() {
        return name;
    }

    public String getUrl() {
        return url;
    }
}

File: BusinessLogic/src/main/java/name/benjaminabbitt/wwwwwh/url/businesslogic/bookmarks/statemachine/states/NotCreatedBookmark.kt
package name.benjaminabbitt.wwwwwh.url.businesslogic.bookmarks.statemachine.states

import com.google.protobuf.Any
import com.google.protobuf.Timestamp
import name.benjaminabbitt.evented.bookmarks.Bookmarks.*
import name.benjaminabbitt.evented.core.Evented
import name.benjaminabbitt.wwwwwh.url.businesslogic.bookmarks.statemachine.InvalidState
import org.slf4j.LoggerFactory
import java.util.stream.Stream

class NotCreatedBookmark : Bookmark(LoggerFactory.getLogger(NotCreatedBookmark::class.java)) {
    var name: String? = null
        private set
    var url: String? = null
        private set

    @Throws(InvalidState::class)
    override fun handle(bookmarkDeleted: BookmarkDeleted?): Bookmark? {
        throw InvalidState()
    }

    @Throws(InvalidState::class)
    override fun handle(deleteBookmark: DeleteBookmark?): Stream<Evented.EventPage.Builder?>? {
        throw InvalidState()
    }

    override fun handle(bookmarkCreated: BookmarkCreated?): Bookmark? {
        url = bookmarkCreated!!.url
        name = bookmarkCreated.name
        return ExtantBookmark(this)
    }

    override fun handle(createBookmark: CreateBookmark?): Stream<Evented.EventPage.Builder?>? {
        return Stream.of(name.benjaminabbitt.evented.core.Evented.EventPage.newBuilder()
                .setCreatedAt(Timestamp.newBuilder().build())
                .setSynchronous(false)
                .setEvent(Any.pack(
                        BookmarkCreated.newBuilder()
                                .setName(createBookmark!!.name)
                                .setUrl(createBookmark.url)
                                .build()
                )))
    }
}